decisions/01-new-animation-api.md

```diff
@@ -0,0 +1,208 @@
+  if (frameCount >= 100) {
+    done(canvas);
+  }
+  // drawing code
+  return canvas;
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-canvas'),
+  mode: 'animation'
+};
+```
+
+```js
+export const handler = async ({ inputs, frame, done, useCanvas }) => {
+  const canvas = useCanvas(inputs.width, inputs.height);
+  let frameCount = 0;
+  const myDrawLoop = () => {
+    frameCount++;
+    // drawing code
+    if (frameCount >= 100) {
+      done(canvas);
+    } else {
+      frame(canvas);
+      window.requestAnimationFrame(myDrawLoop);
+    }
+  };
+  myDrawLoop();
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-canvas'),
+  mode: 'customAnimation'
+};
+```
+
+### `engine-svg`
+
+```js
+export const handler = async ({ inputs, frameCount, done }) => {
+  // drawing code
+  if (frameCount >= 100) {
+    done(svgString);
+  }
+  return svgString;
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-svg'),
+  mode: 'animation'
+};
+```
+
+```js
+export const handler = async ({ inputs, frame, done }) => {
+  let frameCount = 0;
+  const myDrawLoop = () => {
+    frameCount++;
+    // drawing code
+    if (frameCount >= 100) {
+      done(svgString);
+    } else {
+      frame(svgString);
+      window.requestAnimationFrame(myDrawLoop);
+    }
+  };
+  myDrawLoop();
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-svg'),
+  mode: 'customAnimation'
+};
+```
+
+### `engine-react`
+
+```js
+export const handler = async ({ inputs, done, frameCount }) => {
+  if (frameCount >= 100) {
+    done();
+  }
+  return <div></div>;
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-react'),
+  mode: 'animation'
+};
+```
+
+The example above re-renders the component for every frame. If a user wants to use something like `react-spring` to speed up animation while keeping the frame-based approach, it could be done like this:
+
+```js
+export const handler = async ({ inputs, done, frameCount }) => {
+  const spring = useSpring({ frameCount });
+
+  if (frameCount >= 100) {
+    done();
+  }
+  return <MyMemoedComponent spring={spring} />;
+};
+
+const MyMemoedComponent = React.memo(({ spring }) => {
+  return <animated.div></animated.div>;
+});
+
+export const settings = {
+  engine: require('@mechanic-design/engine-react'),
+  mode: 'animation'
+};
+```
+
+```js
+export const handler = async ({ inputs, frame, done }) => {
+  const frameCount = myOwnDrawLoopHook();
+  useEffect(() => {
+    if (frameCount < 100) {
+      frame();
+    } else {
+      done();
+    }
+  }, [frameCount]);
+  return <div></div>;
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-react'),
+  mode: 'customAnimation'
+};
+```
+
+### `engine-p5`
+
+For `engine-p5`, `mode` can only be set to `static` or `customAnimation`. Or perhaps we default both `animation` and `customAnimation` to the code below?
```

Comment:
I like this. Maybe a descriptive info could be printed on the console, to educate developers about why the new animation isn't possible using p5

Reply:
Yes!